### PR TITLE
test: protocol schema CI check; first batch of structs

### PR DESCRIPTION
**Pull request by Longarithm**
Adding "Protocol Schema" check to CI - you can see it in the list of checks. This test will ensure that structs involved in protocol or saved to DB doesn't accidentally change, see #11755. If that's not the case, it will output list of impacted structs and suggest to change `protocol_schema.toml` file.

Example of output in such case:
```
Loaded 123 structs
New struct: FlatStateValue with hash 83834662
New struct: ValueRef with hash 2322946441
New TOML file written to: ./target/protocol_schema.toml
Please review the changes and copy the file to protocol_schema.toml if they are correct.
```

Enabling it for first group of structs, mainly from core/primitives, to check that it works. I will add remaining structs later.

The change is mainly renaming ProtocolStruct with ProtocolSchema as I see it a better fit now, and add ProtocolSchema to the list of derives. Note that if `protocol_schema` and `enable_const_type_id` are not enabled, this derive will be no-op, so it won't impact regular code generation.

cc @nagisa

### CODE REVIEW

File: core/parameters/src/lib.rs [hunk 1]
```diff
@@ -1,3 +1,5 @@
+#![cfg_attr(enable_const_type_id, feature(const_type_id))]
```
Comment (Ekleog-NEAR) [inline core/parameters/src/lib.rs:1]: Why do we need this new line? I could see no other change in this PR that would require this

File: Justfile [hunk 2]
```diff
@@ -159,6 +159,15 @@ check-lychee:
     @echo {{ if env("GITHUB_TOKEN", "") != "" { "" } \
              else { "Note: 'Too Many Requests' errors are allowed here but not in CI, set GITHUB_TOKEN to check them" } }}
 
+# check tools/protocol-schema-check/res/protocol_schema.toml
+check-protocol-schema:
+    rustup toolchain install nightly
+    rustup target add wasm32-unknown-unknown --toolchain nightly
```
Comment (Ekleog-NEAR) [inline Justfile:165]: And in turn if we can remove the feature flag, we will also be able to remove the need for a nightly compiler here

File: core/parameters/src/lib.rs (hunk 1 above)
Comment (Longarithm) [reply core/parameters/src/lib.rs:1]: I need to add this to every crate structs of which I want to guard. This is already in place for near-primitives crate.

The way how I made it work is creating `ProtocolSchemaInfo` for every struct and registering them using `inventory`. To navigate across structs, I also want to record `TypeId`. But `ProtocolSchemaInfo` must be static, and then enabling `const_type_id` becomes a requirement.

Example of resulting struct:
```
pub static WeightedIndex_INFO: near_structs_checker_lib::ProtocolSchemaInfo = near_structs_checker_lib::ProtocolSchemaInfo::Struct {
        name: "WeightedIndex",
        type_id: std::any::TypeId::of::<WeightedIndex>(),
        fields: &[...]
}
```

**Comment by wacban** [comment]
LGTM but I have no context here so I'll leave it to @Ekleog-NEAR to do the review. Please ping me if you'd like my review as well and I'll have a proper look. 

It feels a bit strange that the hashes are numbers and not hex but that's totally fine.

**Comment by Longarithm** [comment]
I can call it a checksum in README. I just needed some simple numbers which will likely be different for different structures, and hashing u32-s seemed the best for that.

File: Justfile (hunk 2 above)
Comment (nagisa) [inline Justfile:165]: I was going to say that this is not the right place to install toolchains or add targets, and that it ought to be done by the calling environment, but then I see that `cargo-udeps` job does the same. So lets leave as-is for now and maybe I'll get around fixing every instance of this later...

File: Justfile [hunk 3]
```diff
@@ -159,6 +159,15 @@ check-lychee:
...
+    # Test that checker is not broken
+    RUSTFLAGS="--cfg enable_const_type_id" cargo +nightly test -p protocol-schema-check
+    # Run the checker
+    RUSTFLAGS="--cfg enable_const_type_id" cargo +nightly run -p protocol-schema-check
```
Comment (nagisa) [inline Justfile:169]:
```suggestion
    env RUSTFLAGS="--cfg enable_const_type_id" cargo +nightly test -p protocol-schema-check
    # Run the checker
    env RUSTFLAGS="--cfg enable_const_type_id" cargo +nightly run -p protocol-schema-check
```

is shell-agnostic and makes copying these commands over to CLI easier.

File: core/parameters/Cargo.toml [hunk 4]
```diff
@@ -26,6 +26,7 @@ clap = { workspace = true, optional = true }
 # FIXME: this ideally should not depend on near-primitives-core...
 near-primitives-core.workspace = true
 near-account-id.workspace = true
+near-structs-checker-lib.workspace = true
```
Comment (nagisa) [inline core/parameters/Cargo.toml:29]: Nit: should the crate be renamed too?

**Review by nagisa — APPROVED** [review]

**Pull request merged**